sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorReceiver.java

```diff
@@ -95,59 +94,73 @@ protected ReactorReceiver(AmqpConnection amqpConnection, String entityPath, Rece
+                                        lastSequenceNumber.set(seqNo);
+                                    }
                                 }
-                            }
 
-                            final int creditsLeft = receiver.getRemoteCredit();
+                                final int creditsLeft = receiver.getRemoteCredit();
+
+                                if (creditsLeft > 0) {
+                                    sink.success(message);
+                                    return;
+                                }
+
+                                final Supplier<Integer> supplier = creditSupplier.get();
+                                final Integer credits = supplier.get();
+
+                                if (credits != null && credits > 0) {
+                                    logger.atInfo()
+                                        .addKeyValue("credits", credits)
+                                        .log("Adding credits.");
+                                    receiver.flow(credits);
+                                } else {
+                                    logger.atVerbose()
+                                        .addKeyValue("credits", credits)
+                                        .log("There are no credits to add.");
+                                }
 
-                            if (creditsLeft > 0) {
+                                metricsProvider.recordAddCredits(credits == null ? 0 : credits);
                                 sink.success(message);
-                                return;
-                            }
-
-                            final Supplier<Integer> supplier = creditSupplier.get();
-                            final Integer credits = supplier.get();
-
-                            if (credits != null && credits > 0) {
-                                logger.atInfo()
-                                    .addKeyValue("credits", credits)
-                                    .log("Adding credits.");
-                                receiver.flow(credits);
-                            } else {
-                                logger.atVerbose()
-                                    .addKeyValue("credits", credits)
-                                    .log("There are no credits to add.");
-                            }
-
-                            metricsProvider.recordAddCredits(credits == null ? 0 : credits);
-                            sink.success(message);
-                        });
-                    } catch (IOException | RejectedExecutionException e) {
-                        sink.error(e);
+                            });
+                        } catch (IOException | RejectedExecutionException e) {
+                            sink.error(e);
+                        }
+                    });
+                }, 1);
+        } else {
+            this.messagesProcessor = this.handler.getDeliveredMessagesV2()
+                .doOnNext(message -> {
```

Comment:
These are supposed to be side effects like logging... since it is part of the chain, should we put this into like a "handle"? Or some other operator?

Reply:
I’ll put this in map, looks like it’s one of the operator with minimal alloc

Reply:
moved to _.map_